### PR TITLE
Add Pull Request template file to the MariaDB/server repository

### PR DESCRIPTION
**Pull request by robertbindar**
I would love for this to be a starting point for a discussion on whether or not we should add a PR template file for the server repository, whether there is a particular format that we should follow if a PR template file is desired, etc.
Basically, I would love to hear your opinions on this.

I made a case in this [comment](https://github.com/MariaDB/server/pull/1776#issuecomment-796783537) for why I believe we definitely need something like this in our daily workflow.

### This is what the contributor needs to edit whilst opening a new PR
Screenshot 2021-03-11 at 23 04 16

### This is how the empty PR template looks like in Preview mode
Screenshot 2021-03-11 at 23 04 22

### This is how the new Pull Request should look like when the sections are filled
Screenshot 2021-03-11 at 23 09 33

### CODE REVIEW

**Comment by robertbindar** [comment]
@ericherman your opinion is highly valuable and appreciated on this ❤️

**Review by grooverdan — COMMENTED** [review]
Good introduction.

Can something early on describe what is the right branch to base/target this pull request against? Users seem to struggle with basing.

Adding to the checklist of "I have created a JIRA task for this Pull Request" I'd include more items checklist like https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/
copied some text and good ideas from https://github.com/embeddedartistry/templates/blob/master/oss_docs/PULL_REQUEST_TEMPLATE.md

## Checklist
- [ ] I have created a JIRA task for this Pull Request
- [ ] My code follows the style guidelines of this project
- [ ] This is the best code I could write and would gladly accept it in my projects
- [ ] I have considered how to ensure this code is of optimal performance
- [ ] I have used portable C/C++/shell code and libraries calls, or have made an effort to ensure that it will work cross platform and OS, using mysys library calls were applicable.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I will try to help MariaDB developers if they, or users, have trouble with this code
- [ ] I have checked my code and corrected any misspellings
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] I have made references in JIRA or here on the corresponding changes to the documentation required
- [ ] I will make changes on the Knowledge base https://mariadb.com/kb/en/documentation/ to ensure that my contribution here is used

**Comment by grooverdan** [comment]
> ## Checklist

Probably note at the bottom of checklist:
 
We understand you have done a lot of work to get this far. We'd like you to consider this checklist and answer honestly.100% ticks are not required for us to consider your contribution.

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+the reviewer reproduce the problem (if applicable).
+-->
+## How can this PR be tested?
+
+<!--
+If applicable. Some contributions only introduce a bug fix
+that is reproducible only on FreeBSD for instance. Please indicate
+if that's the case here.
+-->
+## Please indicate the platform/OS this patch affects
```
[the remark this answers is not in the document]
Comment (robertbindar) [reply .github/pull_request_template.md:61]: I agree, it's probably better to leave this one out and add it later if we see it lacking more often in new PRs.

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+Type of contribution:
+- [ ] Bug fix
+- [ ] New feature (i.e. new functionality that is not expected to affect any other part of the server)
+- [ ] Code refactoring
+- [ ] Documentation
+- [ ] Other
+
+<!--
+An amazing description should answer some questions like:
+1. What problem is the patch trying to solve?
```
[the remark this answers is not in the document]
Comment (robertbindar) [reply .github/pull_request_template.md:41]: I like the issue template you provided above, maybe except the "Describe alternatives" point. I agree some kind of breaking is needed to make sure we get better structured information, but I was afraid not to create a too big of a PR template right from the beginning and scare people, so I chose to rather explain some sub-components in a comment.

[the remark this answers is not in the document]
Comment (robertbindar) [reply .github/pull_request_template.md:41]: About the backwards compatibility section, yeah, I agree, we should think of 1-2-3 critical points that maybe the contributor should check and leave other compatibility considerations up to the reviewer.
I like the two points you mentioned above. Any other ideas?
One addition to the upgrade point you wrote above could be: "Would the server work as expected (crash) if a user starts MariaDB on a datadir created prior to your patch?"

**Comment by robertbindar** [comment]
> > ## Checklist
> 
> Probably note at the bottom of checklist:
> 
> We understand you have done a lot of work to get this far. We'd like you to consider this checklist and answer honestly.100% ticks are not required for us to consider your contribution.

I don't agree that this is the right approach.
Maybe a rephrasing of that would work, but, in the current format, it sounds like we are not offering contributors 100% credibility for the information they provided.
In some rare cases I expect this to be true (i.e. contributors provide wrong information by mistake), but honestly I think we will encounter more cases in which contributors just get scared by the text and erase the PR template text blank before submitting. And even in those mistake cases, reviewer would still have followed the steps in the checklist without the template enabled, so no extra time is wasted compared to now.

**Comment by robertbindar** [comment]
I adjusted the first checkbox of the template according to @vuvova feedback. Also added @grooverdan's suggestions for breaking down the backward compatibility section.

The template should be in a ready to push state already in terms of structure, it's probably a bad idea to add more sections. It's already a bit too big for the first release and I'm afraid it will look scary to contributors.

We should double check the wording is intuitive and english-proof then probably we can push if no more objections arrive.

**Review by iangilfillan — COMMENTED** [review]
Can you clarify how the information here would assist in reviewing the PR? What sort of cases of current PRs are there that waste reviewers time? This solution seems to introduce some duplication.
The first question is about the JIRA issue number, which already appears in more than half of the titles of current PRs, and would also contain the answer to the next question about the type of contribution, and probably later questions too.

**Comment by robertbindar** [comment]
> Can you clarify how the information here would assist in reviewing the PR? What sort of cases of current PRs are there that waste reviewers time? This solution seems to introduce some duplication.
> The first question is about the JIRA issue number, which already appears in more than half of the titles of current PRs, and would also contain the answer to the next question about the type of contribution, and probably later questions too.

Hey @iangilfillan! The PR template is a solution to encourage contributors to include as more information as possible and in the PRs and the benefit of such approach is that a PR template comes almost for free, excepting this .md file we don't have to do anything else. We can also recommend a structure for the information provided which is benefical too.

To be more specific:
- Jira issue number and the type of contribution: yes, probably half the new PRs already include this information (didn't actually check to see what's the proportion), but 50% of PRs that don't include it looks like a lot of work to me. Sure, we won't get a 100% rate suddenly, but even if we get 10% more of new PRs to include the info, it's more compared to the cost of creating such a simple template file.
- I've seen PRs that are created with empty description and even more: empty PR description and oneliner commit message. This creates back and forth messages with the contributor, more context switches for the reviewer, it takes more time to understand the reason of the patch, how to reproduce, etc. These all cost something.
- Put yourself in the shoes of a new contributor coming to the server repository. You see a bug/typo/some compile error, probably you only know how to compile the server to check your patch. You've done the right thing and submitted a patch quickly with your fix. How accessible is for you to know when you create your first PR that you need to e.g. create jira task, add MDEV number in the PR title, keep commit message lines < 80 chars?
This information is not yet quickly available in this repository (thinking again about the CONTRIBUTING.md file), but I would bet that even if such information would be in the repo, lots of contributors wouldn't read it before submitting the first few PRs.

At the end of the day, you can look at this as an experiment, it costs very little to perform, if it works: everybody is happy, if it doesn't: we remove the template from the repo with no negative consequences.

**Comment by robertbindar** [comment]
I would go even further and say:
* you go as a new contributor to jira.mariadb.org and you create an MDEV for a bug
* you select Fix Version: 10.5
* Reviewer sees the PR is made against base branch 10.5, and Jira MDEV specifies FixVersion: 10.5
There is zero guarantees that 10.5 is the earliest server version where the bug can be reproduced, most probably that's the version the contributor has on computer and observed the bug. So Reviewer needs to check 10.4, 10.3.. until bug is not reproducible and then politely ask contributor to rebase the PR to the earliest branch. This takes at least an 1-2 hours of reviewer time.
If the PR template, in at least some cases, encourages contributor to say: "bug reproducible on 10.5, not reproducible on 10.4", that's a huge win in my opinion.

File: .github/pull_request_template.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (grooverdan) [reply .github/pull_request_template.md:40]: > About the backwards compatibility section, yeah, I agree, we should think of 1-2-3 critical points that maybe the contributor should check and leave other compatibility considerations up to the reviewer.
> I like the two points you mentioned above. Any other ideas?

Other things I started to think of became too specific quickly.

> One addition to the upgrade point you wrote above could be: "Would the server work as expected (crash) if a user starts MariaDB on a datadir created prior to your patch?"

I'm all right with the concept, we don't get a lot of on-disk changes however. Careful on wording though.

**Comment by grooverdan** [comment]
> > We understand you have done a lot of work to get this far. We'd like you to consider this checklist and answer honestly.100% ticks are not required for us to consider your contribution.
> 
> I don't agree that this is the right approach.
> Maybe a rephrasing of that would work, but, in the current format, it sounds like we are not offering contributors 100% credibility for the information they provided.

The intent was to encourage the best work possible from the contributors and encourage them if they had nagging doubts on some aspects to rework those, or disclose them. If they really didn't have time, that's ok too, but I'd rather they submit something than be turned away from doing a contribution.

> In some rare cases I expect this to be true (i.e. contributors provide wrong information by mistake), but honestly I think we will encounter more cases in which contributors just get scared by the text and erase the PR template text blank before submitting.

I'd expect mistakes, but lets make them honest mistakes rather than undue pressure to finish.

I suspect by the time someone is ready to submit, they'll erase the template instead of not submitting. I'm happy to see a considered approach that simplifies the questions to a realistic acceptable level. Let's assess the level of acceptance by rates of template erasure.

**Comment by grooverdan** [comment]
Is 10.6 the right branch for this PR :wink:?

Should < 10.6 document have a header that "This is a stable version, the contribution here should be a bug fix, or a minor feature enhancement that has no compatibility issues and doesn't significantly change existing code".

Should the default 10.6 document encourage the setting of the target branch according to the criteria of bug fix.

I'd rather not overly encourage the creation of JIRA issues for spelling mistakes/translation fixes or other really obvious bugs. Our main need is when we need to integrate with the workflow of getting people assigned.

> At the end of the day, you can look at this as an experiment, it costs very little to perform, if it works: everybody is happy, if it doesn't: we remove the template from the repo with no negative consequences.

I'm pretty happy with the PR you've redone it. If you've considered and what to introduce other changes feel free to do so. Lets get something merged to start the experiment.

**Review by an3l — COMMENTED** [review]
Good work Robert, I like this.
I left couple of suggestions regarding the patch and here are some more comments:
- I would say we should have some more information in this form https://github.com/MariaDB/server/pull/1778#pullrequestreview-612773659 or almost everything from that
- Additinally should have a checkbox - > if JIRA MDEV is created (I agree that for some spelling fixes JIRA  is not needed), please make sure that the title of your commit message is the same as JIRA title (or something like that, you get the point)
- Also according to the GitHub documentation https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository
  there may be multiple issue/pull request templates, so please add your .md file to .github/PULL_REQUEST_TEMPLATE/ directory.I think this way you can handle which template to appear according to the different branches (using template query parameter 
  described https://docs.github.com/en/github/managing-your-work-on-github/about-automation-for-issues-and-pull-requests-with-query-parameters)
- Also note that it would be great if you could fill label in GitHub with 10.x when PR is created (I'm not sure is this in the scope of this PR https://github.com/octokit/octokit.net/issues/1862).

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+and have its title be identical to the one of the Pull Request.
+Example:
+"MDEV-21221 Fix crash in sql_acl.cc when new user is created"
+-->
+- [ ] *The jira.mariadb.org issue number for this PR is: MDEV-_____*
+
+<!--
+Please tick the appropriate box by adding an "x" between the square brackets.
+-->
+## Type of contribution
```
[the remark this answers is not in the document]
Comment (cvicentiu) [reply .github/pull_request_template.md:31]: I would not force JIRA onto our contributors as it just increases the difficulty to contribute. Certainly would not force them to create a JIRA entry for their PR if there isn't already one present. We can do that ourselves when dispatching PRs and it's not that big of an overhead.

I am not 100% convinced about the need for this list. The description of "What problem is this patch trying to solve" should make it rather obvious what this is about.

File: .github/pull_request_template.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cvicentiu) [reply .github/pull_request_template.md:40]: I agree with the systemd template from above. I also agree with @robertbindar that the describe alternatives is probably overkill. The main thing we want contributors to do is think how to clearly communicate to us what it is they want to achieve. If we get that through, anything else is trivial.

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+<!--
+You might consider answering some questions like:
+1. Does this affect the on-disk format used by MariaDB?
+2. Does this change any behavior experienced by a user who upgrades from a version prior to this patch?
+-->
+## Backward compatibility
+TODO: fill details here if applicable or remove the section
+
+<!--
```
Comment (cvicentiu) [inline .github/pull_request_template.md:57]: Where does the comment show up for a contributor doing a PR?

Comment (robertbindar) [reply .github/pull_request_template.md:57]: If you look at the screenshots I posted initially, there is an example of how the thing looks when you create a new PR. Basically, the contributor sees a pre-filled PR Description box with the exact content as the template, so the contributor sees the raw markdown text (including comments). Only after the PR is created, the markdown from the template will be rendered (if the contributor chooses not to erase everything, which will definitely happen in some cases 😆  )

File: .github/pull_request_template.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertbindar) [reply .github/pull_request_template.md:31]: You both have some great points on this and I now see it's redundant to have this info in the template. Thanks!

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+The more information you provide about the contribution, the better
+for everybody. You will be happier because your contribution will be
+reviewed faster :-), the reviewer is happier because more Pull Requests
+can be accepted, other contributors are happier because they wait
+less for their PRs, the community is happier because their project
+gets better, faster.
+
+If you have any questions related to MariaDB or you just want to
+hang out and meet other community members, please join us on
+https://mariadb.zulipchat.com/ .
```
[the remark this answers is not in the document]
Comment (robertbindar) [reply .github/pull_request_template.md:15]: I agree it should be there, but let's postpone adding it until we release a facelift version of the CONTRIBUTING doc.

**Comment by robertbindar** [comment]
> Is 10.6 the right branch for this PR 😉?
> 

Good point @grooverdan! Probably we need this in all branches >= 10.2

> Should < 10.6 document have a header that "This is a stable version, the contribution here should be a bug fix, or a minor feature enhancement that has no compatibility issues and doesn't significantly change existing code".
> 
I don't know, but sounds like a good idea, we'll have to add the note every year or so to each branch that becomes stable, but it's a minor cost. I would stick to only "This is a stable version, the contribution here should be a bug fix", "minor feature enhancement" is subjective and I probably do more harm than good.

> Should the default 10.6 document encourage the setting of the target branch according to the criteria of bug fix.
> 
Can you explain this a bit more? I don't understand what you mean here, sorry.

> I'd rather not overly encourage the creation of JIRA issues for spelling mistakes/translation fixes or other really obvious bugs. Our main need is when we need to integrate with the workflow of getting people assigned.
> 
This is a very fine line, I don't think we have a good chance to define it very clearly so that only a small number of contributions fall on one side or the other.

> > At the end of the day, you can look at this as an experiment, it costs very little to perform, if it works: everybody is happy, if it doesn't: we remove the template from the repo with no negative consequences.
> 
> I'm pretty happy with the PR you've redone it. If you've considered and what to introduce other changes feel free to do so. Lets get something merged to start the experiment.

Having small comments/sections to differentiate between (dev, stable) is a great idea in my opinion, feel free to bring more ideas on top of this. As a concrete action, I would first go with having an identical minimal version on all branches and then adjust for each branch type based on ideas like yours from above.

File: .github/pull_request_template.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertbindar) [reply .github/pull_request_template.md:31]: > I would not force JIRA onto our contributors as it just increases the difficulty to contribute. Certainly would not force them to create a JIRA entry for their PR if there isn't already one present. We can do that ourselves when dispatching PRs and it's not that big of an overhead.
> 
This is a great point. Encouraging contributors to create a Jira task themselves would mean creating a new account in Jira for first time contributors which would for sure scale down the enthusiasm for submitting the first contribution. Better leave this one out for now.

> I am not 100% convinced about the need for this list. The description of "What problem is this patch trying to solve" should make it rather obvious what this is about.

File: .github/pull_request_template.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericherman) [reply .github/pull_request_template.md]: I agree as well. If a platform issue has been overlooked, the checkbox reminder is unlikely to illuminate it for the contributor ... if it is platform specific, that's very likely to already be clear in the text of the commit messages.

**Comment by ericherman** [comment]
I think idea is sound.

I would suggest the template could be trimmed down in order to reduce friction. Naturally you're looking to strike a balance ... a little bit of extra work to guide the contributor to get it right the first time is valuable; obviously too much impedes contributions.

I expect some iteration will be required to get something that strikes this balance reasonably well.

Rather than get it perfect the first time, I'd like to see something like this added. Once added, please ensure that at least each of the MariaDB Foundation contributors are prepared *diligently* make use of the template in order to let that guide how it should be improved.

How can it be made more clear with regards to how I should deal with parts of the template which do not make sense in the context of my merge request?

**Comment by robertbindar** [comment]
> I think idea is sound.
> 
> I would suggest the template could be trimmed down in order to reduce friction. Naturally you're looking to strike a balance ... a little bit of extra work to guide the contributor to get it right the first time is valuable; obviously too much impedes contributions.
> 
I agree, especially because the contributor sees the raw markdown when creating the PR and it might be quite hard to follow, resulting in contributor doing Ctrl-A => Backspace.

> I expect some iteration will be required to get something that strikes this balance reasonably well.
> 
> Rather than get it perfect the first time, I'd like to see something like this added. Once added, please ensure that at least each of the MariaDB Foundation contributors are prepared _diligently_ make use of the template in order to let that guide how it should be improved.
> 
Noted.

> How can it be made more clear with regards to how I should deal with parts of the template which do not make sense in the context of my merge request?
My thinking is that the contributor can just erase the section entirely and when the PR is created, that particular section just won't appear in the PR body.
But good point! Ideally there should be a smart and short comment in the template that should quickly help the contributor understand that everything in the template is optional, but the more relevant sections are included, the better for everybody.
Focusing on short and intuitive comments in the template is probably the best solution and we can add a more in-depth guide on how to be a template power user in CONTRIBUTING.md

Many thanks for the comments and ideas @ericherman!

File: .github/pull_request_template.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (an3l) [reply .github/pull_request_template.md:31]: I would suggest to address the comment: https://github.com/MariaDB/server/pull/1778#pullrequestreview-640604785

**Comment by robertbindar** [comment]
Thanks a lot [@grooverdan @cvicentiu @an3l @ericherman @iangilfillan] for the feedback on this. I updated the template based on your comments, it will look something like the picture below when the markdown is rendered.

As a summary:
* comments are now a bit shorter, more details to be added in future `CONTRIBUTING.md` work
* removed `Type of contribution` section
* kept the `Jira Issue Number` checkbox, but adjusted the comment to instruct the contributor to only attach the issue number if contributor is already aware of an existing task tracking the bug/feature implemented (please double check that the wording is intuitive enough)
* Added a one-liner `TODO` for each section where the contributor would have to write text, to help people unfamiliar with markdown not to get lost in the template.
* Added a new section `Basing the PR against the correct MariaDB version` that contains 2 checkboxes.
   The motivation for the two checkboxes is:
   * have a way to instruct the contributors that new features development happen on latest dev branch
   * instruct contributors that, for a bug fix, the PR should be based against the earliest version in which the bug is reproducible
   * achieve the 2 steps above without introducing different versions of the template for (dev, stable) branches
   * make these instructions not sound imperative or like too much work.
Screenshot 2021-04-21 at 16 58 12

**Comment by robertbindar** [comment]
Merged this in 10.6. The PR template file *has* to be in the default branch of the repository, I tested it by pushing in 10.2 and the template wasn't being picked up by GitHub.
